include/aspect/particle/particle.h

```diff
@@ -190,6 +205,23 @@ namespace aspect
         get_location () const;
 
         /**
+         * Set the location of this particle. Note that this does not check
+         * whether this is a valid location in the simulation domain.
+         *
+         * @param [in] new_loc The new location for this particle.
+         */
+        void
+        set_reference_location (const Point<dim> &new_loc);
```

Comment:
You could enforce a simultaneous update if you have just one function that takes both reference and real location. This avoids the possibility of accidentally letting the two get out of synch.